tests/unit/backend/layers/utils/test_spatial_assets_utils.py

```diff
@@ -0,0 +1,251 @@
+
+        self.s3_provider_mock = MagicMock(spec=S3Provider)
+        self.library_id = "abcd123"
+        self.valid_uns = {
+            "spatial": {
+                "library_id_1": {
+                    "images": {"hires": np.random.rand(10, 10, 3), "fullres": np.random.rand(20, 20, 3)},
+                    "scalefactors": {"spot_diameter_fullres": 1.0, "tissue_hires_scalef": 0.5},
+                }
+            }
+        }
+        self.valid_spatial_data = self.valid_uns["spatial"]["library_id_1"]
+        self.invalid_uns = {}
+        self.spatial_processor = SpatialDataProcessor(s3_provider=self.s3_provider_mock)
+        mkdir(self.testing_cxg_temp_directory)
+
+        self.output_folder = "test_output"
+        self.cxg_container = "test_container"
+        self.group_metadata_name = "uns"
+        self.ctx = None
+
+        self.mock_spatial_processor = MagicMock(spec=SpatialDataProcessor)
+        self.mock_spatial_processor.filter_spatial_data.return_value = self.valid_uns["spatial"]
+
+    def test__valid_input_metadata_copy(self):
+        """
+        Test case for verifying the required metadata is present
+        """
+        self.spatial_processor.filter_spatial_data(self.valid_spatial_data, self.library_id)
+
+    def test__invalid_input_metadata_copy(self):
+        """
+        Test case to verify that a KeyError is raised when passing input metadata
+        """
+        with pytest.raises(KeyError):
+            self.spatial_processor.filter_spatial_data(self.invalid_uns, self.library_id)
+
+    def test__full_high_res_image_present(self):
+        """
+        Test that the default image is fullres if present otherwsie fall back to hires
+        """
+        image_array_uint8 = self.spatial_processor._prepare_image(self.valid_spatial_data)
+        self.assertEqual(image_array_uint8.shape, (20, 20, 3))
+
+        valid_hires_uns = copy.deepcopy(self.valid_spatial_data)
+        valid_hires_uns["images"].pop("fullres", None)
+        image_array_uint8 = self.spatial_processor._prepare_image(valid_hires_uns)
+        self.assertEqual(image_array_uint8.shape, (10, 10, 3))
+
+        non_valid_uns = copy.deepcopy(valid_hires_uns)
+        non_valid_uns["images"].pop("hires", None)
+        with pytest.raises(KeyError):
+            image_array_uint8 = self.spatial_processor._prepare_image(non_valid_uns)
+
+    def test__process_and_flip_image(self):
+        """
+        Test the image processing method
+        """
+        test_image_array = (np.random.rand(10, 10, 3) * 255).astype(np.uint8)
+        expected_flipped_array = np.flipud(test_image_array)
+
+        flipped_image_array = self.spatial_processor._process_and_flip_image(test_image_array)
+        assert np.array_equal(flipped_image_array, expected_flipped_array)
+
+    def test__crop_to_aspect_ratio(self):
+        """
+        Test the cropping method to ensure the image is cropped to a square
+        """
+        width, height = 20, 10
```

Comment:
does it also work with odd numbers?

Reply:
yes, I've revised the test to demonstrate that